source/postprocess/visualization/darcy_velocity.cc

```diff
@@ -0,0 +1,135 @@
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/postprocess/visualization/darcy_velocity.h>
+#include <aspect/utilities.h>
+#include <aspect/melt.h>
+#include <aspect/simulator.h>
+
+namespace aspect
+{
+  namespace Postprocess
+  {
+    namespace VisualizationPostprocessors
+    {
+      template <int dim>
+      DarcyVelocity<dim>::
+      DarcyVelocity ()
+        :
+        DataPostprocessorVector<dim> ("darcy_velocity",
+                                      update_values | update_quadrature_points),
+        Interface<dim>()
+      {}
+
+
+
+      template <int dim>
+      std::string
+      DarcyVelocity<dim>::
+      get_physical_units () const
+      {
+        if (this->convert_output_to_years())
+          return "m/year";
+        else
+          return "m/s";
+      }
+
+
+
+      template <int dim>
+      void
+      DarcyVelocity<dim>::
+      evaluate_vector_field(const DataPostprocessorInputs::Vector<dim> &input_data,
+                            std::vector<Vector<double>> &computed_quantities) const
+      {
+        const Quadrature<dim> &quadrature_formula = this->introspection().quadratures.velocities;
+        const unsigned int n_q_points = quadrature_formula.size();
+        AssertThrow(this->introspection().compositional_name_exists("porosity"),
+                    ExcMessage("The 'darcy velocity' postprocessor requires a compositional "
+                               "field called porosity."));
+
+        const unsigned int porosity_idx = this->introspection().find_composition_type(CompositionalFieldDescription::porosity);
+        AssertThrow(this->get_parameters().compositional_field_methods[porosity_idx] ==
+                    Parameters<dim>::AdvectionFieldMethod::fem_darcy_field,
+                    ExcMessage("The 'darcy velocity' postprocessor requires that the compositional "
+                               "field method for 'porosity' is set to 'darcy field'."));
+
+        const double velocity_scaling_factor =
+          this->convert_output_to_years() ? year_in_seconds : 1.0;
+
+        FEValues<dim> fe_values (this->get_mapping(),
+                                 this->get_fe(),
+                                 quadrature_formula,
+                                 update_values |
+                                 update_gradients |
+                                 update_quadrature_points |
+                                 update_JxW_values);
+        std::vector<Tensor<1,dim>> solid_velocities(n_q_points);
+
+        MaterialModel::MaterialModelInputs<dim> in(input_data, this->introspection());
+        MaterialModel::MaterialModelOutputs<dim> out(n_q_points, this->n_compositional_fields());
+        MeltHandler<dim>::create_material_model_outputs(out);
+        this->get_material_model().evaluate(in, out);
+        MaterialModel::MeltOutputs<dim> *fluid_out = out.template get_additional_output<MaterialModel::MeltOutputs<dim>>();
+        AssertThrow(fluid_out != nullptr,
```

Comment:
I do not think this condition is the correct check for this assert. The call to `MeltHandler::create_material_model_outputs` above will always create an object of type MeltOutputs so this check here will always pass. I think what you really want to check is if the material model is a material model that actually fills the object of type MeltOutputs, which unfortunately we do not have a standardized way for. One way to do it would be:
```suggestion
        AssertThrow(std::isfinite(fluid_out->fluid_viscosities[0]),
```
At the moment this does not help much either, because the meltoutputs are default initialized, but I am proposing to change this in #5954